### PR TITLE
Toppings and flavor endpoints

### PR DESCRIPTION
**Pull request by eric-barch**
Addresses Issue #39

- Add MongoDB schema and create router for Toppings
- Populate Toppings collection in db
- Create GET endpoint to retrieve list of available toppings
  - Sample request: ```localhost:3003/api/toppings```
  - Sample response: 
      ```json
      [
        {
          "_id": "6500aab35e28559e2cde18e3",
          "name": "peanuts",
          "imgUrl": "assets/images/toppings/peanuts.png",
          "__v": 0
        },
        {
          "_id": "6500aab45e28559e2cde18e7",
          "name": "hot fudge",
          "imgUrl": "assets/images/toppings/hot-fudge.png",
          "__v": 0
        },
        {
          "_id": "6500aab45e28559e2cde18eb",
          "name": "reeses pieces",
          "imgUrl": "assets/images/toppings/reeses-pieces.png",
          "__v": 0
        },
        // ...15 toppings in total
      ]
      ```
- Create GET endpoint to retrieve x number of flavor suggestions
   - Sample request: ```localhost:3003/api/icecreams/suggestions/2```
   - Response returns 2 ice creams:
      ```json
      [
        {
          "_id": "6500bbaabd5bd2af1ef72c2b",
          "name": "jumbo staircase",
          "isUserGenerated": true,
          "userId": "leewiz",
          "__v": 0
        },
        {
          "_id": "64fb733ee3e373b0b6126544",
          "name": "abominable snowman",
          "isUserGenerated": true,
          "userId": "leewiz",
          "__v": 0
        }
      ]
      ```
    - Sample request asking for more ice creams than exist: ```localhost:3003/api/icecreams/suggestions/7```
    - Response returns all ice creams in database:
       ```json
        [
          {
            "_id": "64fb6f6912f1493fb63d0ddf",
            "name": "seventh",
            "isUserGenerated": true,
            "userId": "dgfbnogrb",
            "__v": 0
          },
          {
            "_id": "64fb733ee3e373b0b6126544",
            "name": "abominable snowman",
            "isUserGenerated": true,
            "userId": "leewiz",
            "__v": 0
          },
          {
            "_id": "6500bbaabd5bd2af1ef72c2b",
            "name": "jumbo staircase",
            "isUserGenerated": true,
            "userId": "leewiz",
            "__v": 0
          }
        ]
       ```

### CODE REVIEW

**Comment by larkinds** [comment]
> Addresses Issue #39
> 
> * Add MongoDB schema and create router for Toppings
> * Populate Toppings collection
> * Create GET endpoint to retrieve list of available toppings
>   
>   * Sample request: `localhost:3003/api/icecreams/`
>   * Sample response:
>     ```json
>     [
>       {
>         "_id": "64fb6f6912f1493fb63d0ddf",
>         "name": "seventh",
>         "isUserGenerated": true,
>         "userId": "dgfbnogrb",
>         "__v": 0
>       },
>       {
>         "_id": "64fb733ee3e373b0b6126544",
>         "name": "abominable snowman",
>         "isUserGenerated": true,
>         "userId": "leewiz",
>         "__v": 0
>       },
>       {
>         "_id": "6500bbaabd5bd2af1ef72c2b",
>         "name": "jumbo staircase",
>         "isUserGenerated": true,
>         "userId": "leewiz",
>         "__v": 0
>       }
>     ]
>     ```
> * Create GET endpoint to retrieve x number of flavor suggestions
>   
>   * Sample request: `localhost:3003/api/icecreams/suggestion/2`
>   * Response returns 2 ice creams:
>     ```json
>     [
>       {
>         "_id": "6500bbaabd5bd2af1ef72c2b",
>         "name": "jumbo staircase",
>         "isUserGenerated": true,
>         "userId": "leewiz",
>         "__v": 0
>       },
>       {
>         "_id": "64fb733ee3e373b0b6126544",
>         "name": "abominable snowman",
>         "isUserGenerated": true,
>         "userId": "leewiz",
>         "__v": 0
>       }
>     ]
>     ```
>   * Sample request asking for more ice creams than exist: `localhost:3003/api/icecreams/suggestion/7`
>   * Response returns all ice creams in database:
>     ```json
>      [
>        {
>          "_id": "64fb6f6912f1493fb63d0ddf",
>          "name": "seventh",
>          "isUserGenerated": true,
>          "userId": "dgfbnogrb",
>          "__v": 0
>        },
>        {
>          "_id": "64fb733ee3e373b0b6126544",
>          "name": "abominable snowman",
>          "isUserGenerated": true,
>          "userId": "leewiz",
>          "__v": 0
>        },
>        {
>          "_id": "6500bbaabd5bd2af1ef72c2b",
>          "name": "jumbo staircase",
>          "isUserGenerated": true,
>          "userId": "leewiz",
>          "__v": 0
>        }
>      ]
>     ```

This is a gorgeous pull request comment, great job Eric & Lee! One comment: the first get request says that it's for toppings, but the sample request and response say icecream.

**Comment by eric-barch** [comment]
Thank you! Got so focused on the formatting that I messed up the content 😅 I updated the comment and also made a few further cleanup changes to the code. Let us know of any further comments!

File: server/controllers/icecreams.ts
Comment (meru-a) [inline server/controllers/icecreams.ts]: Looks like this is unnecessary and could be cleaned up.

Comment (eric-barch) [reply server/controllers/icecreams.ts]: Just the commented line not the whole endpoint right? Agreed, will remove.

**Review by meru-a — COMMENTED** [review]
Great work, @eric-barch and @Leewiz! Very clean and concise code! One thing I wonder if we need - is the `orderItemId` array field inside the `Toppings` schema to link to orders these toppings were used in. I don't see any other way to integrate toppings for now, but please let me know if you think we need this at all? If yes, we might want to add the fields here and in the order item schema in a separate PR.

File: server/controllers/icecreams.ts
Comment (meru-a) [reply server/controllers/icecreams.ts]: Sure, just the line! The endpoint looks perfect otherwise 👍🏼

**Comment by eric-barch** [comment]
> Great work, @eric-barch and @Leewiz! Very clean and concise code! One thing I wonder if we need - is the `orderItemId` array field inside the `Toppings` schema to link to orders these toppings were used in. I don't see any other way to integrate toppings for now, but please let me know if you think we need this at all? If yes, we might want to add the fields here and in the order item schema in a separate PR.

Thanks! And I'm not sure! Usually with a one-to-many relationship like one ```Topping``` to many ```Icecream```/```OrderItem```s,  I would put a field in the "many" record that references the "one" record. So each ```Icecream``` and/or ```OrderItem``` would have a single ```toppingId``` field, rather than ```Topping``` having an array of many ```orderItemIds```. 

It seems like it would be more common to have an ```Icecream``` or ```OrderItem``` and want to get its ```Topping``` than to have a ```Topping``` and want to get a list of *all* of its ```OrderItem```s. But if we anticipate doing the latter a lot then we probably should add an ```orderItemIds``` array.

That's just my first thought and curious what others think. I also feel like I'm not 100% clear on the schema yet and could be misunderstanding.

**Comment by Leewiz** [comment]
i don't remember officially planning a schema for toppings so we went with the simplest approach of getting toppings into the shop. if what was implemented is missing a field required for core features, i'm all for updating it for this PR. otherwise, maybe we hold off until we have a MVP before adding to it.

File: server/requests/icecream-requests.rest
Comment (larkinds) [inline server/requests/icecream-requests.rest]: love the creative icecream name

**Review by larkinds — CHANGES_REQUESTED** [review]
This all looks great, and works perfectly.

One note from me is that the topping's controller doesn't look tested - could you add a file for that with an api route in the requests folder?

**Comment by larkinds** [comment]
I 

> > Great work, @eric-barch and @Leewiz! Very clean and concise code! One thing I wonder if we need - is the `orderItemId` array field inside the `Toppings` schema to link to orders these toppings were used in. I don't see any other way to integrate toppings for now, but please let me know if you think we need this at all? If yes, we might want to add the fields here and in the order item schema in a separate PR.
> 
> Thanks! And I'm not sure! Usually with a one-to-many relationship like one `Topping` to many `Icecream`/`OrderItem`s, I would put a field in the "many" record that references the "one" record. So each `Icecream` and/or `OrderItem` would have a single `toppingId` field, rather than `Topping` having an array of many `orderItemIds`.
> 
> It seems like it would be more common to have an `Icecream` or `OrderItem` and want to get its `Topping` than to have a `Topping` and want to get a list of _all_ of its `OrderItem`s. But if we anticipate doing the latter a lot then we probably should add an `orderItemIds` array.
> 
> That's just my first thought and curious what others think. I also feel like I'm not 100% clear on the schema yet and could be misunderstanding.

I agree - seems like 

> i don't remember officially planning a schema for toppings so we went with the simplest approach of getting toppings into the shop. if what was implemented is missing a field required for core features, i'm all for updating it for this PR. otherwise, maybe we hold off until we have a MVP before adding to it.

Totally - 

> i don't remember officially planning a schema for toppings so we went with the simplest approach of getting toppings into the shop. if what was implemented is missing a field required for core features, i'm all for updating it for this PR. otherwise, maybe we hold off until we have a MVP before adding to it.

Agreed - it's not in the schema. I feel fine about holding off on this update for the sake of the MVP - we don't care about tracking toppings at this point, since they don't show up in the hall of fame page anyways

**Comment by meru-a** [comment]
> Thanks! And I'm not sure! Usually with a one-to-many relationship like one `Topping` to many `Icecream`/`OrderItem`s, I would put a field in the "many" record that references the "one" record. So each `Icecream` and/or `OrderItem` would have a single `toppingId` field, rather than `Topping` having an array of many `orderItemIds`.
> 
> It seems like it would be more common to have an `Icecream` or `OrderItem` and want to get its `Topping` than to have a `Topping` and want to get a list of _all_ of its `OrderItem`s. But if we anticipate doing the latter a lot then we probably should add an `orderItemIds` array.

Makes sense! In MongoDB terms, we could have both fields and make them update each other if necessary (which totally might not be). I also agree with Lee's point that for the sake of MVP we can skip this integration for now :)

File: server/requests/icecream-requests.rest
Comment (meru-a) [inline server/requests/icecream-requests.rest]: Forgot to mention - I was testing the endpoint yesterday and here had to change for `suggestions` to launch the request. But everything worked like a charm!

**Review by meru-a — APPROVED** [review]

File: server/requests/icecream-requests.rest
Comment (eric-barch) [reply server/requests/icecream-requests.rest]: thanks! this is now updated in the latest on this branch

**Comment by eric-barch** [comment]
> This all looks great, and works perfectly.
> 
> One note from me is that the topping's controller doesn't look tested - could you add a file for that with an api route in the requests folder?

thanks! we added a test for toppings in the requests directory

**Comment by eric-barch** [comment]
I think this is ready to merge into staging unless anyone finds new issues. 

- Added a test GET request for the toppings endpoint in ```topping-requests.rest```
- Found and resolved a bug that was causing the server to crash when a non-unique icecream POST request was made.  The solution was to remove ```.send()``` from the end of the response command, as the ```.json()``` method chained before it already sends a response.
  - Example: ```response.status(200).json(icecream).send();``` -> ```response.status(200).json(icecream);```
  - Made this change globally to all controllers to hopefullly avoid future bugs
  - Tested all endpoints after making the change - everything still works and sends expected responses.
- Resolved conflicts with staging

**Review by larkinds — APPROVED** [review]
Oooh great catch with the .json().send() bug. Well done, approved

**Pull request merged**
